### PR TITLE
Fix ChunkedCELoss final FSDP wrapping

### PR DESCRIPTION
**Pull request by iamzainhuda**
## Summary

- keep final `norm` and `lm_head` in separate FSDP units when `ChunkedCELoss` is active
- avoid forcing an immediate `lm_head` reshard at the end of the chunked loss forward
- replace the custom chunked-loss autograd bridge with an equivalent zero-value gradient bridge
- add unit coverage showing the zero-value bridge returns the same scalar and computes the same hidden-state gradients as the legacy custom autograd bridge

`ChunkedCELoss` runs backward through `lm_head` chunk-by-chunk before the outer decoder backward reaches final `norm`. When `norm` and `lm_head` share one FSDP unit, `lm_head` backward/reshard can affect storage that `norm` still needs later in the same backward pass. On the 8-GPU DeepSeek V3 debug model this fails during the first backward pass with an unallocated tensor error.

Splitting `norm` and `lm_head` only for the chunked-loss path isolates their FSDP storage lifecycle while preserving the existing grouped wrapping for the standard loss path.

The zero-value gradient bridge keeps the same forward value as the old custom autograd Function while attaching the accumulated chunk gradient to `hidden_states` through regular PyTorch ops. The added unit test recreates the legacy bridge locally and verifies both approaches produce identical scalar values and hidden-state gradients.

## Validation

- `PYTHONPATH=/home/dev/ao pytest -q tests/unit_tests/test_loss.py`
- latest `origin/main` fails with `NGPU=8 MODULE=deepseek_v3 CONFIG=deepseek_v3_debugmodel ./run_train.sh --training.steps 3`
- this branch completes the same 8-GPU DeepSeek V3 debug-model smoke for 3 steps

### CODE REVIEW

**Comment by wwwjn** [comment]
> latest origin/main fails with NGPU=8 MODULE=deepseek_v3 CONFIG=deepseek_v3_debugmodel ./run_train.sh --training.steps 3

Hi thanks for this. Did you run this test with latest pytorch nightly? We have CI to  guard this which is passing on ChunkedLoss change. ChunkedLoss depends on PyTorch changes https://github.com/pytorch/pytorch/pull/180428

**Comment by iamzainhuda** [comment]
The custom autograd function change is an additional thing I added that isn't actually needed to make this work but thought we could simplify. I'm happy to remove it if you want to keep it in. There is a small transient memory tradeoff because `hidden_states * accumulated_grad`, but this isn't persistent mem.

**Comment by iamzainhuda** [comment]
> > latest origin/main fails with NGPU=8 MODULE=deepseek_v3 CONFIG=deepseek_v3_debugmodel ./run_train.sh --training.steps 3
> 
> Hi thanks for this. Did you run this test with latest pytorch nightly? We have CI to guard this which is passing on ChunkedLoss change. ChunkedLoss depends on PyTorch changes [pytorch/pytorch#180428](https://github.com/pytorch/pytorch/pull/180428)

Just ran it on the latest pytorch nightly, the issue is gone - so this PR is no longer needed. Thank you for the response!
